keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,347 @@
+* Validation of `VolumeSource` will ensure that only `CSIVolumeSource` is being copied.  
+* External CSI attacher must be extended to  process either `PersistentVolumeName` or `VolumeSource`.
+* **External attacher may need permission to access secrets in any pod namespace** where inline volume is specified.
+* CSI `ControllerUnpublishVolume` call (~ volume detach) will require the secrets to be available at detach time. 
+* If user deletes secrets in a pod's namespace that was used to attach an inline volume, the external attacher will fail during detach (volume will remain attached), reporting errors about missing secrets to user.
+
+For persistent inline drivers:
+* Attachment will require a volumeHandle
+* As stated, volumeHandle must refer to a pre-existing provisioned volume
+
+For ephemeral inline drivers:
+* Ephemeral inline drivers must ignore attachment requests
+* Attachment requires a volumeHandle value which could be generated at that stage by Kubernetes
+* For security reason, autogenerated volumeHandle should include podUID info which is not available at this stage
+
+### Mount/Unmount
+This phase happens in the Kubelet and is responsible for mounting/unmounting device and/or filesystem mount points.  At this stage, volume operations have access to pod information such as podUID.  The volume information will come from `volume.Spec` which contains either `v1.CSIVolumeSource` (for volume originated from pod specs) or `v1.CSIPersistentVolume` for volume originating from PV/PVC.
+
+In-tree CSI volume plugin calls in kubelet, get universal `volume.Spec`, which contains either `v1.VolumeSource` from Pod (for inline volumes) or `v1.PersistentVolume` (originated from PV/PVC). The code will check for presence of `CSIVolumeSource` or `CSIPersistentVolume` and read NodeStage/NodePublish secrets from appropriate source. Kubelet does not need any new permissions, it already can read secrets for pods that it handles. These secrets are needed only for `MountDevice/SetUp` calls and don't need to be cached until `TearDown`/`UnmountDevice`.
+
+For persistent inline drivers:
+
+* The Kubelet code will follow the natural volume operation phases supported by CSI PV/PVC volumes
+* The Kubelet code will extract volume information from `v1.CSIVolumeSource` including `volumeHandle`.
+* Kubelet will delegate storage operations (mount, unmount, etc) to external CSI driver calls respectively
+
+
+For ephemeral inline drivers:
+* During the Setup/Mount stage, podUID will be available
+* The Kubelet will create necessary mount point paths
+* Kubelet will auto-generate a volumeHandle based on podUID (see above for detail)
+* Ephemeral drivers will receive mount-like calls (NodeStage/NodePublish) with generated paths and volumeHandle
```

Comment:
Node*Stage won't be supported. Only Node*Publish.

Reply:
Not supporting `NodeStageVolume` will force existing drivers which currently rely on it to do some extra work in `NodePublishVolume`. Based on my current understanding of the non-goals that's fine, I just wanted to clarify.

Reply:
There are no ephemeral drivers today. There is no use case envisioned where an ephermal driver would need both NodeStage and NodePublish. It is also much more complicated to make NodeStage available in the code as it is missing some information that is already in NodePublish. Drivers will already need to be modified to work as ephemeral drivers since the volume provisioning/deprovisioning needs to happen during publish. So I think limiting it to only Publish/unpublish is a good tradeoff.

Reply:
Second @kfox1111 . This would be only ephemeral drivers which don't exist today.

Reply:
I don't disagree and came to the same conclusion following the same logic. I just wanted to clarify that drivers indeed need to be modified for this to work.